### ISSUE TITLE
Issue `/save-all` on shudown

### ISSUE THREAD

**Issue by spannerman79** [Gamealition/ForgeAutoShutdown#6]
Just before mod induced shutdown issue a `/save-all` within the mod coding with enough time to ensure data is fully written to `world/`

**Comment by RoyCurtis** [comment]
This is a good idea. Initially, I was going to reject and close on the understanding that the shutdown process implicitly does a `/save-all`. However, upon walking through the code I realized it would be more reliable for ForgeAutoShutdown to issue a `/save-all` as you suggest.

So, thank you for the idea. I will not be implementing it (I don't want to go through the testing and packaging process) but would recommend any future maintainers do so.

Since I've written half of this post already, I am going to go ahead and finish it and share my findings :wink: I will walk through the shutdown mechanism using Forge's decompilation of Minecraft.

----

When shutting down, [ForgeAutoShutdown will kick all players](https://github.com/Gamealition/ForgeAutoShutdown/blob/master/1.8/src/main/java/roycurtis/autoshutdown/util/Server.java#L20) and then execute `MinecraftServer.initiateShutdown()`.
* This is always done in the server thread, whilst the server is handling events, to ensure stability
* The method of using `initiateShutdown()` was figured out by [looking at what the `/stop` command does, internally.](http://vanderprot.gamealition.com/img/76cbe.png)

If we look into what `initiateShutdown()` does, [we can see it merely sets a boolean `serverRunning` to false](http://vanderprot.gamealition.com/img/9664f.png). If we follow the usages, we find the main loop of the server; [a `while` loop that checks `serverRunning` is true](http://vanderprot.gamealition.com/img/064b3.png).

If we assume `serverRunning` is false and continue past the while loop (ignoring `handleServerStopping()` and `expectServerStopped()`, since these are what looks like Forge handlers), [there's eventually a `finally` block that executes `stopServer()`.](http://vanderprot.gamealition.com/img/f88c7.png)

Going into `stopServer()`, a lot of things appear to be handled. [One of them is `saveAllWorlds(false)`](http://vanderprot.gamealition.com/img/3bf82.png)\*

If we continue into `saveAllWorlds()`, we can see that all loaded worlds have their chunks saved. A shutdown doesn't show log messages because of `false` passed on to `saveAllWorlds()`, but it does indeed save.

So in conclusion, the vanilla shutdown process (e.g. `/stop` or ForgeAutoShutdown) does save all worlds, but it would be more reliable to issue a `/save-all` first anyway. I think the way to do this is to [look at what the command does](http://vanderprot.gamealition.com/img/a6e21.png) and make sure to do it within the server thread.

----

\* At this point, I realized that [if anything goes wrong between](http://vanderprot.gamealition.com/img/eb692.png) `serverRunning` being set to false and `saveAllWorlds(false)` being called, a shutdown (via `/stop` or ForgeAutoShutdown) could very well indeed fail to save any data.

**Comment by spannerman79** [comment]
I'm still learning and trying to understand coding but I had this "gut feeling" that if something does go wrong during a shutdown data would be lost.  Not sure if I saw something in the coding or knew for some other reason.  Your actual walk through actually shines a light on my actual gut feeling so thank you for taking the time for writing your actual explanation :smile: 

I was looking for a mod that does do a timed shutdown as I already have a restart shell script in place for what i dub "the kids server" which is a vanilla forge server.  KCauldron is able to do what your mod does via a plugin (I used [ConsoleScheduler](http://dev.bukkit.org/bukkit-plugins/consolescheduler2/) with correctly timed `/save-all` and `/stop`) but for other reasons I cannot have a KCauldron server running for the kids to connect to - especially when they have learning difficulties and have troubles trying to understand what they did wasn't their fault but because of a bug with the server ([this](https://gitlab.prok.pw/KCauldron/KCauldron/issues/275) as an example)
